### PR TITLE
Loss going NaN because of log(0)

### PR DESCRIPTION
**Pull request by p-sodmann**
My loss went to NaN because somewhere a log(y_pred) was log(0)

to prevent this, I added a small value (epsilon = 1e-5) that fixes that issue.
The loss works on one-dimensional data as well btw.

### CODE REVIEW

File: losses/binary_losses.py [hunk 1]
```diff
@@ -44,7 +45,7 @@ def convert_to_logits(y_pred: tf.Tensor) -> tf.Tensor:
     # To avoid unwanted behaviour of log operation
     y_pred = K.clip(y_pred, tf.keras.backend.epsilon(), 1 - tf.keras.backend.epsilon())
 
-    return K.log(y_pred / (1 - y_pred))
+    return K.log(y_pred / (1 - y_pred) + epsilon)
```
Comment (maxvfischer) [inline losses/binary_losses.py:48]: Is this epsilon really necessary? What I can tell, the clipping in the line above should handle so that `y_pred != 0` and `y_pred != 1`

File: losses/binary_losses.py [hunk 2]
```diff
@@ -245,7 +246,7 @@ def loss(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
         :param y_pred: Predicted masks (tf.Tensor, shape=(<BATCH_SIZE>, <IMAGE_HEIGHT>, <IMAGE_WIDTH>, 1))
         :return: Focal loss (tf.Tensor, shape=(<BATCH_SIZE>,))
         """
-        f_loss = beta * (1 - y_pred) ** gamma * y_true * K.log(y_pred)  # β*(1-p̂)ᵞ*p*log(p̂)
+        f_loss = beta * (1 - y_pred) ** gamma * y_true * K.log(y_pred + epsilon)  # β*(1-p̂)ᵞ*p*log(p̂)
```
Comment (maxvfischer) [inline losses/binary_losses.py:249]: This makes sense. What do you think about putting a `K.clip(y_pred, tf.keras.backend.epsilon(), 1 - tf.keras.backend.epsilon())` above instead?

File: losses/multiclass_losses.py [hunk 3]
```diff
@@ -136,7 +137,7 @@ def loss(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
         y_pred = K.clip(y_pred, K.epsilon(), 1-K.epsilon())  # To avoid unwanted behaviour in K.log(y_pred)
 
         # p * log(p̂) * class_weights
-        wce_loss = y_true * K.log(y_pred) * class_weights
+        wce_loss = y_true * K.log(y_pred + epsilon) * class_weights
```
Comment (maxvfischer) [inline losses/multiclass_losses.py:140]: Same here, to what I can see, this should be handled by the clipping in the line above? Or?

File: losses/multiclass_losses.py [hunk 4]
```diff
@@ -173,7 +174,7 @@ def loss(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
         :param y_pred: Predicted masks (tf.Tensor, shape=(<BATCH_SIZE>, <IMAGE_HEIGHT>, <IMAGE_WIDTH>, <N_CLASSES>))
         :return: Focal loss (tf.Tensor, shape=(None,))
         """
-        f_loss = -(class_weights * (1-y_pred)**gamma * y_true * K.log(y_pred))
+        f_loss = -(class_weights * (1-y_pred)**gamma * y_true * K.log(y_pred + epsilon))
```
Comment (maxvfischer) [inline losses/multiclass_losses.py:177]: This makes sense. Maybe add a `K.clip(y_pred, tf.keras.backend.epsilon(), 1 - tf.keras.backend.epsilon())` instead of epsilon?

**Comment by p-sodmann** [comment]
Yes clipping would probably be better, since it does not always change the output.
Good catch.
